autopush-common/src/db/bigtable/bigtable_client/mod.rs

```diff
@@ -193,6 +194,85 @@ fn to_string(value: Vec<u8>, name: &str) -> Result<String, DbError> {
     })
 }
 
+pub fn retry_policy(max: usize) -> RetryPolicy {
+    RetryPolicy::default()
+        .with_max_retries(max)
+        .with_jitter(true)
+}
+
+fn retriable_internal_error(status: &RpcStatus) -> bool {
+    match status.code() {
+        RpcStatusCode::UNKNOWN => {
+            "error occurred when fetching oauth2 token" == status.message().to_ascii_lowercase()
+        }
+        RpcStatusCode::INTERNAL => [
+            "rst_stream",
+            "rst stream",
+            "received unexpected eos on data from from server",
+        ]
+        .contains(&status.message().to_lowercase().as_str()),
+        RpcStatusCode::UNAVAILABLE => true,
+        RpcStatusCode::DEADLINE_EXCEEDED => true,
+        _ => false,
+    }
+}
+
+pub fn retryable_describe_table_error(
+    metrics: Arc<StatsdClient>,
+) -> impl Fn(&grpcio::Error) -> bool {
+    move |err| {
+        debug!("🉑 Checking error...{err}");
+        match err {
+            grpcio::Error::RpcFailure(status) => {
+                info!("GRPC Failure :{:?}", status);
+                metrics
+                    .incr_with_tags("database.retry")
+                    .with_tag("error", "RpcFailure")
+                    .with_tag("type", "bigtable")
+                    .with_tag("code", &status.code().to_string())
+                    .send();
+                retriable_internal_error(status)
+            }
+            grpcio::Error::QueueShutdown => {
+                metrics
+                    .incr_with_tags("database.retry")
+                    .with_tag("error", "QueueShutdown")
+                    .with_tag("type", "bigtable")
+                    .send();
+                true
+            }
+            grpcio::Error::BindFail(_) => {
+                metrics
+                    .incr_with_tags("database.retry")
+                    .with_tag("error", "BindFail")
+                    .with_tag("type", "bigtable")
+                    .send();
+                true
+            }
+            // The parameter here is a [grpcio_sys::grpc_call_error] enum
+            // Not all of these are retriable.
+            grpcio::Error::CallFailure(grpc_call_status) => {
+                metrics
+                    .incr_with_tags("database.retry")
+                    .with_tag("error", "CallFailure")
+                    .with_tag("type", "bigtable")
+                    .with_tag("code", &format!("{:?}", grpc_call_status))
+                    .send();
+                grpc_call_status == &grpcio_sys::grpc_call_error::GRPC_CALL_ERROR
```

Comment:
Where did you copy these retry checks from, I thought all I saw in the Python client was basically the `RpcFailure` block above (I'm not sure what `QueueShutdown` or `ShutdownFailed` are but their description makes them sound like they shouldn't be retried).

Reply:
A combination of the python and java code. I'm fine not calling those, so I'll remove them. 

Both were a bit vague, partly because if the shutdown succeeds, then the next request goes to a different source or fails for a different reason. These appeared to be more warning messages.